File: packages/swt3-ai/swt3_ai/chain.py
"""SWT3 AI Witness SDK -- Client-side chain verification.

Walk provenance reference chains using locally available data.
No HTTP calls -- the caller provides a lookup map built from their payloads.
"""

from __future__ import annotations

from collections import deque
from typing import Any, Dict, List, Optional, Sequence, Tuple

from .types import AnchorReference, ChainLink, ChainSummary


def build_lookup(
    payloads: Sequence[Dict[str, Any]],
) -> Dict[str, List[AnchorReference]]:
    """Build a lookup dict from an array of payload-like dicts.

    Each dict must have an ``anchor_fingerprint`` key and optional ``references`` list.
    References can be dicts with a ``fingerprint`` key or AnchorReference instances.
    """
    result: Dict[str, List[AnchorReference]] = {}
    for p in payloads:
        fp = p.get("anchor_fingerprint", "")
        refs_raw = p.get("references") or []
        refs: List[AnchorReference] = []
        for r in refs_raw:
            if isinstance(r, AnchorReference):
                refs.append(r)
            elif isinstance(r, dict):
                refs.append(AnchorReference(
                    fingerprint=r.get("fingerprint", ""),
                    relationship=r.get("relationship"),
                    provenance_token=r.get("provenance_token"),
                ))
            else:
                refs.append(AnchorReference(fingerprint=str(r)))
        result[fp] = refs
    return result


def walk_chain(
    start_fingerprint: str,
    lookup: Dict[str, List[AnchorReference]],
    max_depth: int = 10,
) -> ChainSummary:
    """Walk a provenance chain starting from a fingerprint using BFS.

    Cycle-safe via visited set. Depth-limited via max_depth.
    """
    links: List[ChainLink] = []
    gaps: List[str] = []
    visited: set = set()
    max_reached = 0
    truncated = False

    queue: deque[Tuple[str, int]] = deque()
    queue.append((start_fingerprint, 0))

    while queue:
        fp, depth = queue.popleft()

        if fp in visited:
            continue
        visited.add(fp)

        if depth > max_depth:
            truncated = True
            continue

        refs = lookup.get(fp)
        if refs is None:
            gaps.append(fp)
            continue

        if depth > max_reached:
            max_reached = depth

        links.append(ChainLink(fingerprint=fp, references=list(refs), depth=depth))

        for ref in refs:
            if ref.fingerprint not in visited:
                queue.append((ref.fingerprint, depth + 1))

    return ChainSummary(
        root=start_fingerprint,
        links=links,
        depth=max_reached,
        gaps=gaps,
        complete=len(gaps) == 0,
        truncated=truncated,
    )


def verify_chain_integrity(chain: ChainSummary) -> Dict[str, Any]:
    """Verify the integrity of a chain walk result.

    Returns dict with ``intact`` (bool) and ``issues`` (list of strings).
    """
    issues: List[str] = []

    if not chain.links:
        issues.append("Chain has no resolved links")

    if chain.gaps:
        issues.append(f"{len(chain.gaps)} referenced fingerprint(s) not found in lookup")

    if chain.truncated:
        issues.append("Chain traversal was truncated at max depth")

    return {"intact": len(issues) == 0, "issues": issues}

File: packages/swt3-ai-ts/src/chain.ts
/**
 * SWT3 AI Witness SDK -- Client-side chain verification.
 *
 * Walk provenance reference chains using locally available data.
 * No HTTP calls -- the caller provides a lookup map built from their payloads.
 */

import type { AnchorReference, ChainLink, ChainSummary } from "./types.js";

/**
 * Build a lookup map from an array of payload-like objects.
 * Each object must have an `anchor_fingerprint` and optional `references` array.
 */
export function buildLookup(
  payloads: Array<{ anchor_fingerprint: string; references?: AnchorReference[] }>,
): Map<string, AnchorReference[]> {
  const map = new Map<string, AnchorReference[]>();
  for (const p of payloads) {
    if (p.references && p.references.length > 0) {
      map.set(p.anchor_fingerprint, p.references);
    } else {
      map.set(p.anchor_fingerprint, []);
    }
  }
  return map;
}

/**
 * Walk a provenance chain starting from a fingerprint using BFS.
 * Cycle-safe via visited set. Depth-limited via maxDepth.
 *
 * @param startFingerprint - The anchor to start from.
 * @param lookup - Map from fingerprint to its references (from buildLookup).
 * @param maxDepth - Maximum traversal depth (default 10).
 */
export function walkChain(
  startFingerprint: string,
  lookup: Map<string, AnchorReference[]>,
  maxDepth = 10,
): ChainSummary {
  const links: ChainLink[] = [];
  const gaps: string[] = [];
  const visited = new Set<string>();
  let maxReached = 0;
  let truncated = false;

  // BFS queue: [fingerprint, depth]
  const queue: Array<[string, number]> = [[startFingerprint, 0]];

  while (queue.length > 0) {
    const [fp, depth] = queue.shift()!;

    if (visited.has(fp)) continue;
    visited.add(fp);

    if (depth > maxDepth) {
      truncated = true;
      continue;
    }

    const refs = lookup.get(fp);
    if (refs === undefined) {
      if (fp !== startFingerprint) {
        gaps.push(fp);
      } else {
        // Start fingerprint not in lookup -- it's still a gap
        gaps.push(fp);
      }
      continue;
    }

    if (depth > maxReached) maxReached = depth;
    links.push({ fingerprint: fp, references: refs, depth });

    for (const ref of refs) {
      if (!visited.has(ref.fingerprint)) {
        queue.push([ref.fingerprint, depth + 1]);
      }
    }
  }

  return {
    root: startFingerprint,
    links,
    depth: maxReached,
    gaps,
    complete: gaps.length === 0,
    truncated,
  };
}

/**
 * Verify the integrity of a chain walk result.
 * Returns whether the chain is intact and any issues found.
 */
export function verifyChainIntegrity(chain: ChainSummary): { intact: boolean; issues: string[] } {
  const issues: string[] = [];

  if (chain.links.length === 0) {
    issues.push("Chain has no resolved links");
  }

  if (chain.gaps.length > 0) {
    issues.push(`${chain.gaps.length} referenced fingerprint(s) not found in lookup`);
  }

  if (chain.truncated) {
    issues.push("Chain traversal was truncated at max depth");
  }

  return { intact: issues.length === 0, issues };
}
